### PR TITLE
tests/e2e: Introduce new e2e operator test

### PR DESCRIPTION
**Pull request by kfirtoledo**
Add a new e2e test for checking operator deployment.

### CODE REVIEW

File: tests/e2e/k8s/test_operator.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	"github.com/clusterlink-net/clusterlink/pkg/bootstrap/platform"
+	"github.com/clusterlink-net/clusterlink/pkg/operator/controller"
+	"github.com/clusterlink-net/clusterlink/tests/e2e/k8s/services"
+	"github.com/clusterlink-net/clusterlink/tests/e2e/k8s/services/httpecho"
+	"github.com/clusterlink-net/clusterlink/tests/e2e/k8s/util"
+)
+
+// TestOperator test the operator functionality.
+func (s *TestSuite) TestOperator() {
+	s.RunSubTest("TestOperator", func() {
```
Comment (elevran) [inline tests/e2e/k8s/test_operator.go:37]: Is there a value in having a single SubTest in a Suite as opposed to individual `TestXYZ` functions?

**Review by elevran — APPROVED** [review]

File: tests/e2e/k8s/test_operator.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+package k8s
+
+import (
+	"context"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	apimachinerywait "k8s.io/apimachinery/pkg/util/wait"
+
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:22]:
```suggestion
```

File: tests/e2e/k8s/test_operator.go (hunk 1 above)
Comment (orozery) [reply tests/e2e/k8s/test_operator.go:37]: +1 You should move the body of this function straight to `TestOperator`.

File: tests/e2e/k8s/util/fabric.go [hunk 3]
```diff
@@ -310,6 +311,16 @@ func (f *Fabric) DeployClusterlinks(peerCount uint8, cfg *PeerConfig) ([]*Cluste
 	return clusterlinks, nil
 }
 
+// GetPeerResources returns peer resources.
+func (f *Fabric) GetPeerResources(num int) *resources.Resources {
```
Comment (orozery) [inline tests/e2e/k8s/util/fabric.go:315]: I would define `GetPeerCluster(num int) *util.KindCluster` instead.

File: tests/e2e/k8s/util/fabric.go [hunk 4]
```diff
@@ -310,6 +311,16 @@ func (f *Fabric) DeployClusterlinks(peerCount uint8, cfg *PeerConfig) ([]*Cluste
...
+	return f.peers[num].cluster.resources
+}
+
+// GetNamespace returns fabric namespace.
+func (f *Fabric) GetNamespace() string {
```
Comment (orozery) [inline tests/e2e/k8s/util/fabric.go:320]:
```suggestion
func (f *Fabric) Namespace() string {
```

File: tests/e2e/k8s/test_operator.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+				Namespace:         s.fabric.GetNamespace(),
+				ContainerRegistry: "docker.io/library", // Tell kind to use local image.
+				Ingress:           clusterlink.IngressSpec{Type: "NodePort", Port: int32(cl[0].Port())},
+			},
+		}
+
+		peerResource := s.fabric.GetPeerResources(0)
+		err = peerResource.Create(context.Background(), instance2)
+
+		// Check basic connectivity with operator
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:63]:
```suggestion
		// Check basic connectivity with instances deployed by operator
```

File: tests/e2e/k8s/test_operator.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		require.Nil(s.T(), cl[1].CreateImport("echo", importedService))
+		require.Nil(s.T(), cl[1].CreateBinding("echo", cl[0]))
+		require.Nil(s.T(), cl[1].CreatePolicy(util.PolicyAllowAll))
+
+		data, err := cl[1].AccessService(httpecho.GetEchoValue, importedService, true, nil)
+		require.Nil(s.T(), err)
+		require.Equal(s.T(), cl[0].Name(), data)
+
+		// Check instance1 deployment succeed
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:81]:
```suggestion
		// Check that instance1 deployment succeeded
```

File: tests/e2e/k8s/test_operator.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		instance1 := &clusterlink.Instance{}
+		err = peerResource.Get(context.Background(), "cl-instance"+s.fabric.GetNamespace(), controller.OperatorNamespace, instance1)
+		require.Nil(s.T(), err)
+		require.Equal(s.T(), v1.ConditionTrue, instance1.Status.Controlplane.Conditions[string(clusterlink.DeploymentReady)].Status)
+		require.Equal(s.T(), v1.ConditionTrue, instance1.Status.Dataplane.Conditions[string(clusterlink.DeploymentReady)].Status)
+		require.Equal(s.T(), v1.ConditionTrue, instance1.Status.Ingress.Conditions[string(clusterlink.ServiceReady)].Status)
+
+		// Check instance2 fails
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:89]:
```suggestion
		// Verify instance2 fails
```

File: tests/e2e/k8s/test_operator.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		err = peerResource.Get(context.Background(), "instance2", controller.OperatorNamespace, instance2)
+		require.Nil(s.T(), err)
+		require.Equal(s.T(), v1.ConditionFalse, instance2.Status.Controlplane.Conditions[string(clusterlink.DeploymentReady)].Status)
+		require.Equal(s.T(), v1.ConditionFalse, instance2.Status.Dataplane.Conditions[string(clusterlink.DeploymentReady)].Status)
+
+		// Delete first instance.
+		err = peerResource.Delete(context.Background(), instance1)
+		require.Nil(s.T(), err)
+
+		// Check instance2 succeed.
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:102]:
```suggestion
		// Check that instance2 succeeded.
```

File: tests/e2e/k8s/test_operator.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		require.Nil(s.T(), cl[0].CreateExport("echo", &httpEchoService))
+		require.Nil(s.T(), cl[0].CreatePolicy(util.PolicyAllowAll))
+		data, err = cl[1].AccessService(httpecho.GetEchoValue, importedService, true, nil)
+		require.Nil(s.T(), err)
+		require.Equal(s.T(), cl[0].Name(), data)
+	})
+}
+
+// instanceCPConditionMatch is a helper function that can be used to check a specific condition match in the controlplane status.
+func instanceCPConditionMatch(
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:123]: Can this function be inlined (perhaps even getting rid of the arguments which are always the same) to the test function?

**Review by orozery — CHANGES_REQUESTED** [review]

File: tests/e2e/k8s/test_operator.go (hunk 9 above)
Comment (kfirtoledo) [reply tests/e2e/k8s/test_operator.go:123]: But I called it twice, I prefer not to duplicate the code in this case.

File: tests/e2e/k8s/util/fabric.go (hunk 3 above)
Comment (kfirtoledo) [reply tests/e2e/k8s/util/fabric.go:315]: But the resources field is not exposed.

Comment (orozery) [reply tests/e2e/k8s/util/fabric.go:315]: Add `Resources()` function to `KindCluster`.

File: tests/e2e/k8s/test_operator.go (hunk 9 above)
Comment (orozery) [reply tests/e2e/k8s/test_operator.go:121]: That's not what I meant :)
I meant moving the function definition inside the test function. e.g.
`instanceCPConditionMatch := func(...) {}`
Then you can get rid of 3 of its 4 arguments which are fixed.

**Comment by kfirtoledo** [comment]
@orozery PTAL

File: tests/e2e/k8s/test_operator.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	// Check that instance1 deployment succeeded.
+	instance1 := &clusterlink.Instance{}
+	err = peerResource.Get(context.Background(), "cl-instance"+s.fabric.Namespace(), controller.OperatorNamespace, instance1)
+	require.Nil(s.T(), err)
+	require.Equal(s.T(), v1.ConditionTrue, instance1.Status.Controlplane.Conditions[string(clusterlink.DeploymentReady)].Status)
+	require.Equal(s.T(), v1.ConditionTrue, instance1.Status.Dataplane.Conditions[string(clusterlink.DeploymentReady)].Status)
+	require.Equal(s.T(), v1.ConditionTrue, instance1.Status.Ingress.Conditions[string(clusterlink.ServiceReady)].Status)
+
+	// Verify that instance2 failed.
+	instanceCPConditionMatch := func(conditionStatus v1.ConditionStatus) apimachinerywait.ConditionWithContextFunc {
```
Comment (orozery) [inline tests/e2e/k8s/test_operator.go:87]:
```suggestion
	instance2ReadyCondition := func(conditionStatus v1.ConditionStatus) apimachinerywait.ConditionWithContextFunc {
```

**Review by orozery — APPROVED** [review]

**Pull request merged**